### PR TITLE
Add support for user-defined settings

### PR DESCRIPTION
**Pull request by shabalind**
This change makes settings an open set via a `BenchmarkSetting` protocol, replacing the previous enum-based `BenchmarkSetting` which could not be amended by the users.

In addition, command-line arguments are factored out into their own arguments bundle called `BenchmarkArguments`. The bundle can be easily reused to define custom CLI commands that also include benchmark arguments (without having to redefine them).

To add new setting(s), `swift-benchmark` user needs to:

1) Define a new struct(s) for the setting that conforms to `BenchmarkSetting`.
2) Define a custom `ParseableCommand` that includes `BenchmarkArguments`, and add any extra arguments in there.
3) Define a helper that converts command arguments into `BenchmarkSetting`-s, similar to the one defined in `BenchmarkArguments`. 
4) Use a new main overload that accepts a sequence of settings, and allows to use custom command parser beforehand. 
5) (optional) define helpers that simplify access on `BenchmarkSettings` container, similarly to the one provided for the default settings.

### CODE REVIEW

File: Sources/Benchmark/BenchmarkRunner.swift [hunk 1]
```diff
@@ -80,7 +81,7 @@ public struct BenchmarkRunner {
     func predictNumberOfIterationsNeeded(_ measurements: [Double], settings: BenchmarkSettings)
         -> Int
     {
-        let minTime = settings.minTime
+        let minTime = settings.minTime!
```
Comment (saeta) [inline Sources/Benchmark/BenchmarkRunner.swift]: Why the change, and would this cause problems in some circumstances?

File: Sources/Benchmark/BenchmarkSetting.swift [hunk 2]
```diff
@@ -12,68 +12,106 @@
...
-    init(
-        iterations: Int?, warmupIterations: Int?, maxIterations: Int, filter: String?,
-        minTime: Double
-    ) throws {
-        self.iterations = iterations
-        self.warmupIterations = warmupIterations
-        self.maxIterations = maxIterations
-        self.filter = try BenchmarkFilter(filter)
-        self.minTime = minTime
+    public var minTime: Double? {
```
Comment (saeta) [inline Sources/Benchmark/BenchmarkSetting.swift]: Can you think of a circumstance where `minTime` should be allowed to be `nil`? If not, could we move this validation into the initializer, which will thus be much closer to the actual bug?

File: Sources/Benchmark/BenchmarkSetting.swift [hunk 3]
```diff
@@ -12,68 +12,106 @@
...
+        }
+
+        self.init(result)
+    }
+
+    public init() {
+        self.init([:])
+    }
+
+    init(_ settings: [String: BenchmarkSetting]) {
```
Comment (saeta) [inline Sources/Benchmark/BenchmarkSetting.swift:92]: Why not use the metatypes themselves instead of a string representation of them?

File: Sources/Benchmark/BenchmarkRunner.swift [hunk 4]
```diff
@@ -103,16 +104,16 @@ public struct BenchmarkRunner {
...
-        let nextIters = min(maxNextIters, settings.maxIterations)
+        let nextIters = min(maxNextIters, settings.maxIterations!)
 
         return nextIters
     }
 
     /// Heuristic when to stop looking for new number of iterations, ported from google/benchmark.
     func hasCollectedEnoughData(_ measurements: [Double], settings: BenchmarkSettings) -> Bool {
-        let tooManyIterations = measurements.count >= settings.maxIterations
+        let tooManyIterations = measurements.count >= settings.maxIterations!
```
Comment (saeta) [inline Sources/Benchmark/BenchmarkRunner.swift]: Ditto here too. (See related comment below.)

**Review by saeta — APPROVED** [review]

File: Sources/Benchmark/BenchmarkRunner.swift (hunk 1 above)
Comment (shabalind) [reply Sources/Benchmark/BenchmarkRunner.swift]: Refactored in 30e4368

File: Sources/Benchmark/BenchmarkRunner.swift (hunk 4 above)
Comment (shabalind) [reply Sources/Benchmark/BenchmarkRunner.swift]: Refactored in 30e4368

File: Sources/Benchmark/BenchmarkSetting.swift (hunk 3 above)
Comment (shabalind) [reply Sources/Benchmark/BenchmarkSetting.swift:92]: We use type name because `Any.Type` is not Hashable by default. If we had support for private/scoped conformances we could define one that uses type name, but otherwise it seems risky to provide it globally.

File: Sources/Benchmark/BenchmarkSetting.swift (hunk 2 above)
Comment (shabalind) [reply Sources/Benchmark/BenchmarkSetting.swift]: Refactored in 30e4368

**Comment by shabalind** [comment]
I'm closing this one for now (and it's LGTM-approved already) to unblock dependant and conflicting PRs.
